planner/core/rule_predicate_simplification.go

```diff
@@ -0,0 +1,157 @@
+	"context"
+
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/parser/ast"
+)
+
+// predicateSimplification consolidates different predcicates on a column and its equivalence classes.  Initial out is for
+// in-list and not equal list intersection.
+type predicateSimplification struct {
+}
+
+type predicateType = byte
+
+const (
+	inListPredicate   predicateType = 0x00
+	notEqualPredicate predicateType = 0x01
+	otherPredicate    predicateType = 0x02
+)
+
+func findPredicateType(expr expression.Expression) (*expression.Column, predicateType) {
+	switch v := expr.(type) {
+	case *expression.ScalarFunction:
+		args := v.GetArgs()
+		col, colOk := args[0].(*expression.Column)
+		if !colOk {
+			return nil, otherPredicate
+		}
+		if v.FuncName.L == ast.NE {
+			if _, ok := args[1].(*expression.Constant); !ok {
+				return nil, otherPredicate
+			}
+			return col, notEqualPredicate
+		}
+		if v.FuncName.L == ast.In {
+			for _, value := range args[1:] {
+				if _, ok := value.(*expression.Constant); !ok {
+					return nil, otherPredicate
+				}
+			}
+			return col, inListPredicate
+		}
+	default:
+		return nil, otherPredicate
+	}
+	return nil, otherPredicate
+}
+
+func (s *predicateSimplification) optimize(_ context.Context, p LogicalPlan, opt *logicalOptimizeOp) (LogicalPlan, error) {
+	return p.predicateSimplification(opt), nil
+}
+
+func (s *baseLogicalPlan) predicateSimplification(opt *logicalOptimizeOp) LogicalPlan {
+	p := s.self
+	for i, child := range p.Children() {
+		newChild := child.predicateSimplification(opt)
+		p.SetChild(i, newChild)
+	}
+	return p
+}
+
+func updateInPredicate(inPredicate expression.Expression, notEQPredicate expression.Expression) expression.Expression {
+	_, inPredicateType := findPredicateType(inPredicate)
+	_, notEQPredicateType := findPredicateType(notEQPredicate)
+	if inPredicateType != inListPredicate || notEQPredicateType != notEqualPredicate {
+		return inPredicate
+	}
+	v := inPredicate.(*expression.ScalarFunction)
+	notEQValue := notEQPredicate.(*expression.ScalarFunction).GetArgs()[1].(*expression.Constant)
+	newValues := make([]expression.Expression, 0, len(v.GetArgs()))
+	var lastValue *expression.Constant
+	for _, element := range v.GetArgs() {
+		value, valueOK := element.(*expression.Constant)
+		redudantValue := valueOK && value.Equal(v.GetCtx(), notEQValue)
+		if !redudantValue {
+			newValues = append(newValues, element)
+		}
+		if valueOK {
+			lastValue = value
+		}
+	}
+	// Special case if all IN list values are prunned. Ideally, this is False condition
+	// which can be optimized with LogicalDual. But, this ia already done. TODO: the false
+	// optimization and its propagation through query tree will be added part of predicate simplification.
+	if len(newValues) < 2 {
+		newValues = append(newValues, lastValue)
+	}
+	newPred := expression.NewFunctionInternal(v.GetCtx(), v.FuncName.L, v.RetType, newValues...)
+	return newPred
+}
+
+func indexInSlice(index int, list []int) bool {
```

Comment:
You can use `slices.Contains` by importing `golang.org/x/exp/slices`.

Reply:
Sounds good. Will be done in next round.